README.md
# bbc-xfer
Transfer files between a BBC Micro and PC. A popular, easy-to-use command-line utility written by Mark de Weger, ported to C by Angus Duggan and extended by Jon Welch.

This version includes some enhancements added to accomodate personal requirements:

* Continue on ADFS disk transfer errors (disk read, CRC, timeout etc). This is a WIP. In future this will be switchable and I will also do the same for DFS.
* Force ADFS single-sided transfer even if disc is detected as double-sided (command 'E'). This is useful if the disk is accidently formatted as DFS on what would be drive 2.


docs/changes.txt
$Id: Changes.txt,v 1.2 2005/01/22 16:47:52 jon Exp $

XFer in C Version 5.3a
======================

  * Correctly crunched XFER.BAS

XFer in C Version 5.3
=====================

  * Compiled for 32-bit Windows, Raspbian Wheezy and i386 Linux
  * Linux backend imported from version 4.0
  * Skips TAPE keyword in INF files, corrected NEXT syntax

XFer in C Version 5.2 - Created by Greg Cook - debounce@yahoo.co.uk
=====================

  * Enabled CTS/RTS handshake mode
  * Improved INF handling - default attributes if INF file empty,
    length optional, length checked against file if present,
    CRC and NEXT attributes skipped
  * Help displayed on request, so more dialogue stays on screen
  * Fixed checksumming of PC files
  * Removed delays in the middle of sending data buffers to the BBC
  * Crunched XFER.BAS by 70 percent

XFer in C Version 5.1
=====================

  * Support for writing DFS and ADFS disc images back to BBC
  * Show PC current directory on menu prompt

XFer in C Version 5.0 - Created by Jon Welch - jon.welch@ntlworld.com
=====================

  * Support for ADFS disc image downloading
  * Corrected bugs in DFS disc image downloading
  * Support for DFS double sided dsd disc image downloading

XFer in C Version 4.0
=====================

- Complete rewrite for Win32 (Windows 95/98/NT/2000) and Linux in C
  * Open source
  * Support for disc image downloading
  * Use of OSFILE to read file attributes
  * Extended .inf file format for file attributes
  * Improved CRC support
  * Support for recursive directory downloading
  * Support for multi-format disc filing systems (HDFS)

XFer Version 3.0
================

- Program now handles disk boot option (*OPT 4) parameter
  * When transferring an !BOOT file from a BBC to a PC, the disk
    boot option parameter is added to the !BOOT.INF file.
  * When transferring an !BOOT file from a PC to a BBC, while the
    !BOOT.INF file contains a boot option parameter, the correct
    boot option is set on the BBC disk.
  * See file FORMAT.TXT for explanation of syntax of !BOOT.INF.
  * (Note: boot option is transferred for every !BOOT file, even
    "non-bootable" !BOOT files, such as !BOOT files in directories
    other than $.)


XFer Version 2.3
================

- Bug fix
  * program crashing when transferring files with length of exact
    multiple of 4,000 bytes (4,000 bytes is BBC buffer size)
- More documentation on troubleshooting.


XFer Version 2.2
================

- Bug fix
  * program unable to transfer first file in BBC *INFO list if option
    "wildcards" turned on.
- More documentation on troubleshooting.


XFer Version 2.1
================

- Bug fix
  * program crashing when trying to transfer files from empty disks.
- Program renamed to XFer.


BBCXFer Version 2
=================

- Bug fixes:
  * program crashing when transferring 0-byte files;
  * no re-checking for existing file name (to prevent overwriting existing
    files) after incoming file has already been renamed.
- Improved instructions for making program work with non-standard DFSs
  and BBC Master.
- Added instructions on how to prevent problems when running BBCXFer
  under Windows95.
- Program confirmed to work with Acorn ADFS, many other DDFSs, and
  on BBC Master.


BBCXFer Version 1 (First Public Release)
========================================

No modifications


BBCXFer version Beta 2
======================

- Bug fixes:
  * terminal emulator crashing in very specific circumstances;
  * program deadlocking after sending 17 files in a row to the BBC
    (was a pretty hard one!);
  * receiving files impossible without option "wildcards" set to off
    with standard Acorn DFS (it should only be necessary to set the option
    "wildcards" to off when using a DFS that is not compatible with a
    standard Acorn DFS).
- Improved instructions for making serial cable in case 25 pin PC port
  is used.
- Modified name of "image format for storing BBC files on other computers"
  to "standard format for storing BBC files on other computers" in all
  documents.
- Improved description of standard format for storing BBC files on other
  computers.


BBCXFer version Beta 1
======================

First beta version


src/xfer.h
#ifndef XFER_H
#define XFER_H
/* $Id: xfer.h,v 1.10  2005/01/16 10:47:53  jon Exp $
 *
 * Interface between main program and serial routines for XFer in C
 *
 * Copyright (C) Mark de Weger 1997, Angus Duggan 1999, Jon Welch 2005,
 *               Greg Cook 2013
 *
 * Additional modifications done by Jon Welch - jon.welch@ntlworld.com
 * Additional modifications done by Greg Cook - debounce@yahoo.co.uk
 *
 * Freely redistributable software. See file LICENSE for details.
 */

/* Version of XFer in C. Initial version is 4.0 to distinguish from original
   XFer, which reached version 3.0 */
#define XFER_VERSION 5.3

#include <stdarg.h>
#ifdef _WIN32
#include <direct.h>
#define mkdir(_p, _m) _mkdir(_p)
#define chdir(_p) _chdir(_p)
#define getcwd(_p, _l) _getcwd(_p, _l)
extern void sleep(int seconds) ;
#else /* _WIN32 */
#include <sys/stat.h>
#include <sys/types.h>
#include <fcntl.h>
#include <unistd.h> /* getcwd, chdir, mkdir, sleep */
#endif /* _WIN32 */

#define RESTART_LINE "PROCmain"
#define TEXT_SYNC    "-----BBC-----PC-----"
#define TEXT_SYNC_LEN 20
#define ERR_TXT      "-----BBCerror1-----PC-----"
#define ERR_TXT_LEN 26
#define ERR_TXT2     "-----BBCerror2-----PC-----"
#define ERR_TXT2_LEN 26

#define MAXLINELEN 256          /* Max length for BBC commands, BBC/PC filenames */
#define BBCTRACKSIZE 2560       /* Number of bytes in a standard BBC track */
#define BBCADFSTRACKSIZE 4096   /* Number of bytes in a standard ADFS BBC track */
#define SENDBUFSIZE 4096        /* This *must* match bufsize% in xfer.bas */

/* Protocol will be incremented if there are changes in the protocol which
   required a change at the BBC end. The initial value is chosen to be higher
   than any baud rate that the BBC supports, to distinguish it from older
   versions of the XFer basic program. */
#define XFER_PROTOCOL_VERSION 100001

void error(char *format, ...) ;

typedef enum {false, true} bool ;

typedef enum {
  XFER_B1200, XFER_B2400, XFER_B4800, XFER_B9600, XFER_B19200, N_BAUD_RATES
} baud_rate_t ;

/* Handshake methods. Only CTS/RTS and DSR/DTR are currently supported. */
typedef enum {
  HANDSHAKE_XON_XOFF, HANDSHAKE_CTS_RTS, HANDSHAKE_DSR_DTR
} handshake_t ;

typedef struct { int dummy ; } *serial_h ;
typedef struct { int dummy ; } *dir_h ;

typedef enum {
  SERIAL_OK, SERIAL_TIMEOUT, SERIAL_ERROR
} serial_status_t ;

typedef enum {
  BBC_OK, BBC_ERROR, BBC_ERROR_2, BBC_SYNC
} bbc_status_t ;

typedef struct {
  char *comport, *basic ;
  bool wildcards, directories ;
  baud_rate_t baudrate ;
  handshake_t handshake ;
  long timeoutdelay ;
} configuration_t ;

/* Serial port management */
serial_status_t serial_open(configuration_t *config, serial_h *connection) ;
serial_status_t serial_config(serial_h connection, configuration_t *config) ;
serial_status_t serial_write(serial_h connection, char *buffer, long size) ;
serial_status_t serial_flush(serial_h connection) ;
serial_status_t serial_read(serial_h connection, char *buffer, long *size) ;
void serial_close(serial_h connection) ;

/* Formatted serial transfer routine with error handling */
void serial_printf(serial_h connection, char *format, ...) ;

/* Line and block-reading functions */
bbc_status_t bbc_readline(serial_h connection, char *buffer, long size) ;
bbc_status_t bbc_read(serial_h connection, char *buffer, long *size) ;

/* Directory searching */
bool forallfiles(char *pattern, bool (*fn)(char *filename, void *data),
                 void *data) ;

/* Configuration management */
void default_configuration(configuration_t *config, char *home) ;

/* First time connection functions */
bool first_time_ready(void) ;
bool first_time_send(serial_h com, char *file, bool autorun) ;

/* Utility functions */
bool itobaud(long baud, baud_rate_t *value) ;
long baudtoi(baud_rate_t value) ;
bool strtobool(char *str, bool *value) ;
bool strtohandshake(char *str, handshake_t *value) ;
bool match(char *pattern, char *string) ;

/* Transfer dialogue */
void xfer(serial_h com, configuration_t *config) ;

/* Keyboard control */
void keyboard_open(void) ;
char keyboard_char(void) ;
long keyboard_line(char *buffer, long size) ;
void keyboard_close(void) ;

/* Interrupt handling */
void interrupt_catch(void) ;
bool interrupted(bool reset) ;

/*
 * $Log: xfer.h,v $
 * Revision 1.11  2005/01/22 08:27:39  jon
 * Added support for transferring disc images back from PC to BBC
 *
 * Revision 1.10  2005/01/16 10:47:53  jon
 * Added ADFS support
 *
 * Revision 1.9  1999/11/08 09:42:48  angus
 * Win32 interface for sleep()
 *
 * Revision 1.8  1999/11/08 09:38:34  angus
 * Urgh. Revolting changes to get send file to avoid buffer overruns. Sending at
 * slow speed may be the only way to prevent this...
 *
 * Revision 1.7  1999/11/07 23:32:55  angus
 * Allow quitting before connected
 *
 * Revision 1.6  1999/11/06 23:04:20  angus
 * Add local and remote checksums, correct messages, quit from disc image
 * retrieve, subroutines for filespecs, dot file fixes.
 *
 * Revision 1.5  1999/11/05 02:43:05  angus
 * Linux changes; timeouts for serial read, keyboard open/close, directory
 * searching for send file.
 *
 * Revision 1.4  1999/11/04 09:21:16  angus
 * Many more changes. Altered some protocols, added text files, renumber script,
 * rewrite machine code, add track return, send file, match files, etc.
 *
 * Revision 1.3  1999/11/02 08:36:46  angus
 * Many updates; Win32 keyboard/console/serial interface, serial transfer, info
 * file read and save.
 *
 * Revision 1.2  1999/10/27 07:01:11  angus
 * Add main.c and first.c, split first-time transfer, parameter
 * encapsulation, initial configuration, argument parsing
 *
 * Revision 1.1.1.1  1999/10/25 16:51:23  angus
 * Development CVS import
 *
 */
#endif /* XFER_H */
